data/src/main/java/com/microsoft/azure/kusto/data/ClientDetails.java

```diff
@@ -0,0 +1,166 @@
+package com.microsoft.azure.kusto.data;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.concurrent.ConcurrentException;
+import org.apache.commons.lang3.concurrent.ConcurrentInitializer;
+import org.apache.commons.lang3.concurrent.LazyInitializer;
+import org.apache.commons.lang3.tuple.Pair;
+import reactor.util.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class ClientDetails {
+
+    public static final String NONE = "[none]";
+
+    private static String unpackLazy(ConcurrentInitializer<String> s) {
+        try {
+            return s.get();
+        } catch (ConcurrentException e) {
+            return NONE;
+        }
+    }
+
+    private static ConcurrentInitializer<String> defaultApplication = new LazyInitializer<>() {
+        @Override
+        protected String initialize() {
+            return UriUtils.stripFileNameFromCommandLine(System.getProperty("sun.java.command"));
+        }
+    };
+
+    private static ConcurrentInitializer<String> defaultUser = new LazyInitializer<>() {
+        @Override
+        protected String initialize() {
+            String user = System.getProperty("user.name");
+            if (StringUtils.isBlank(user)) {
+                user = System.getenv("USERNAME");
+                String domain = System.getenv("USERDOMAIN");
+                if (StringUtils.isNotBlank(domain) && StringUtils.isNotBlank(user)) {
+                    user = domain + "\\" + user;
+                }
+            }
+            return StringUtils.isNotBlank(user) ? user : "[none]";
+        }
+    };
+
+    private static ConcurrentInitializer<String> defaultVersion = new LazyInitializer<>() {
+        @Override
+        protected String initialize() {
+            return formatHeader(Arrays.asList(
+                    Pair.of("Kusto.Java.Client", Utils.getPackageVersion()),
+                    Pair.of(String.format("Runtime.%s", escapeField(getRuntime())), getJavaVersion())));
+        }
+    };
+
+    /**
+     * Formats the given fields into a string that can be used as a header.
+     *
+     * @param args The fields to format.
+     * @return The formatted string, for example: "field1:{value1}|field2:{value2}"
+     */
+    private static String formatHeader(Collection<Pair<String, String>> args) {
+        return args.stream().filter(arg -> org.apache.commons.lang3.StringUtils.isNotBlank(arg.getKey()) && StringUtils.isNotBlank(arg.getValue()))
+                .map(arg -> String.format("%s:%s", arg.getKey(), escapeField(arg.getValue())))
+                .collect(Collectors.joining("|"));
+    }
+
+    private static String escapeField(String field) {
+        return String.format("{%s}", field.replaceAll("[\\r\\n\\s{}|]+", "_"));
```

Comment:
Why the whole `String.format("{%s}",`?
Don't you just want `field.replaceAll("[\\r\\n\\s{}|]+", "_")`?

Reply:
The brackets are literal. For example "123" -> "{123}"

Reply:
My mistake. I must be losing my ability to support multiple languages at once :-D